src/rpc/net.cpp

```diff
@@ -185,6 +185,14 @@ static RPCHelpMan getpeerinfo()
         UniValue obj(UniValue::VOBJ);
         CNodeStateStats statestats;
         bool fStateStats = peerman.GetNodeStateStats(stats.nodeid, statestats);
+        // GetNodeStateStats() requires the existence of a CNodeState and a Peer object
+        // to succeed for this peer. These are created at connection initialisation and
+        // exist for the duration of the connection - except if there is a race where the
+        // peer got disconnected in between the GetNodeStats() and the GetNodeStateStats()
+        // calls. In this case, the peer doesn't need to be reported here.
+        if (!fStateStats) {
```

Comment:
Haven't tested yet if this works, but checking for `fStateStats` seems a bit of a smell and I wanted to remove it entirely, e.g. https://github.com/bitcoin/bitcoin/pull/25923.

Note that you haven't updated the code for the GUI peers details.

Reply:
Looking into #25923 right now - it seems that it does some what Marco suggested below, making some fields, like txrelay, optional for an intrinsic reason instead of using defaults like -1 or 0 when the info is not known.

As for the GUI - I didn't change it on purpose, because [this comment](https://github.com/bitcoin/bitcoin/blob/6863ad79a65842504ab6f5584fac3d1de7ecf87e/src/qt/rpcconsole.cpp#L1200-L1201) scared me off. I think that #25923 is affected by that too, I'll comment there.

Reply:
The `TRY_LOCK` for accessing `CNodeStats` has been in the GUI since peer information was added in #4225 (see https://github.com/bitcoin/bitcoin/pull/4225#issuecomment-44150459). I'm not sure why it was decided that the GUI could block on taking `cs_vNodes`, but can't block on taking `cs_main`.